arraycontext/container/traversal.py

```diff
@@ -179,12 +179,13 @@ def map_array_container(
         or an instance of a base array type.
     """
     try:
-        ser_ctr = serialize_container(ary)
-    except TypeError:
+        iterable = serialize_container(ary)
+    except NotImplementedError:
```

Comment:
Is the `TypeError` needed here?

Reply:
The object array serialization can raise `ValueError` for non-object arrays. Is it worth catching that here?

Reply:
So a `ValueError` to me indicates specifically that whatever was passed as `ary` is simply not acceptable. I guess one thing to ask here is what is the formal definition of an "array container". A `NotImplementedError` to me indicates that whatever I passed is an "array container", but is just missing an implementation. So I think if something like a float (I guess a scalar element?) comes in, it should trigger a `ValueError`; floats are not "array containers"

Any thoughts? This is just my mental picture of all this.

Reply:
Currently the `serialize_container` / `deserialize_container` act more like methods on an abstract base class and `is_array_container` just checks if they're implemented for the given type (using [singledispatch.dispatch](https://docs.python.org/3/library/functools.html#functools.singledispatch)).

The default implementation just raises `NotImplementedError` to say "this type does not implement the ABC" and I'm not sure raising `ValueError` makes more sense here. I'm also becoming less convinced that the `ndarray` serialization should raise `NotImplementedError` for other dtypes, since that would cause a mismatch with what `is_array_container` says..

Probably an issue for another day, i.e. #28.

Reply:
I'm leaning towards sticking with `TypeError`, actually.

* Why not `ValueError`? That says to me: "the type you're passing is OK, but the specific value/instances of the type you chose is terrible."
* Why not `NotImplementedError`? Agree with @thomasgibson here that that says "eh, we haven't gotten around to it yet." Which may have made sense in the early days of array containers when there were plenty of things that probably should be containers that weren't, but that's no longer today's reality.

> I'm also becoming less convinced that the ndarray serialization should raise `NotImplementedError` for other dtypes, since that would cause a mismatch with what `is_array_container` says.

I think `is_array_container` is just wrong in that case. I'm pretty convinced that plain-old numpy arrays aren't containers. I'd be more inclined to redefine `is_array_container` as "try to serialize it, and it's not a container if that fails". We're on that road with this PR anyway.

Reply:
I.e. what I mean here is to change the default implementation of `{de,}serialize_container` to raise `TypeError`.

Reply:
Done in baaab74.

Agreed that `is_array_container` not quite doing its job and we should just let `serialize_container` figure out if it can do it or not.